### DISCUSSION TITLE
v0.6.0 Pre-release thread

### DISCUSSION THREAD

**Discussion by masenf** [reflex-dev/reflex#3912] [Release]
# Try out the Prerelease

### If you are a **third party component developer or have complex wrapped components**, _please test your code against the 0.6.0 pre-release as **semantics of `rx.Var` have changed significantly**_.

Please report any problems you encounter in this discussion thread or as separate issues on reflex-dev/reflex.

```console
pip install 'reflex~=0.6.0a'
```

# Schedule

2024-09-11 13:00 PDT: 0.6.0a1 pre-release published for testing
2024-09-17 11:45 PDT: 0.6.0a2 pre-release published for testing
2024-09-20 10:00 PDT: 0.6.0a3 pre-release published
2024-09-24 08:30 PDT: 0.6.0a4 pre-release published
2024-09-24 2:45 PDT: 0.6.0 final release

# Changes in 0.6.0a4

* Bring back py3.9 support with a deprecation warning. (#3976)
* [ENG-3833] handle object in is bool (#3974)
* suggest bool() for wrong values (#3975)
* use is true for bool var (#3973)
* fix set value logix for client state (#3966)

# Changes in 0.6.0a3

* bump python>=3.10 for 0.6.0 (#3956)
* use current version as default for new custom components (#3957)
* Add shim for `format_event_chain` (#3958)
* use serializer before serializing base yourself (#3960)
* [ENG-3817] deprecate _var_name_unwrapped (instead of removing it) (#3951)
* move the filterwarning to appropriate file (#3952)
* fix unionize recursion (#3948)
* add special handling for infinity and nan (#3943)
* use is true (#3946)

# Changes in 0.6.0a2

* [0.6.0 blocker] state: update inherited_vars and tracking dicts when adding vars (#2822)
* expose radix primitive progress under rx.radix.primitives.progress (#3930)
* use serializer for state update and rework serializers (#3934)
* better errors in state.py (#3929)
* fix template fetch during init (#3932)
* Fix string color (#3922)
* replace old var system with immutable one (#3916)
* Remove Pydantic from some classes (#3907)
* [REF-3562][REF-3563] Replace chakra usage (#3872)

# Release Notes

### Drop support for <del>python 3.8 and</del> python3.9
* bump python>=3.10 for 0.6.0 (#3956)

### Remove Deprecated Features
* `rx.input.root`
* ... ⚠️

* [REF-3568][REF-3569]Remove deprecations by @ElijahAhianyo in https://github.com/reflex-dev/reflex/pull/3852
* [REF-3570] Remove deprecated REDIS_URL syntax by @masenf in https://github.com/reflex-dev/reflex/pull/3892

### New Var System

A `Var` is a placeholder for a value which is sent to the frontend or rendered as Javascript code in the app. Most users of Reflex do not interact with the var system directly, but it becomes relevant for advanced users that are wrapping components or handling custom types.

* `rx.Var.create` now returns an `ImmutableVar` or `LiteralVar` subclass wrapping the provided value
* `_var_is_local` - removed, use `rx.ivars.ImmutableVar.create`
* `_var_is_string` - removed, passing a string will always create a `StringVar`
* Vars are not automatically cast as `bool`, when passing a non-bool var to a prop that expects a boolean, add `.bool()` after the var.

The new `Var` system brings a much more accurate translation of literal python values into javascript, especially of object and array types.

### `rx.chakra` removed from main repo

First, `pip install reflex-chakra`. Then `import reflex_chakra as rc`. `rx.chakra` will still work until 0.7.0, but it is recommended to move now to avoid deprecation warning.

* Upper bound for reflex-chakra dependency by @masenf in https://github.com/reflex-dev/reflex/pull/3824
* [REF-3566]Remove demo template by @ElijahAhianyo in https://github.com/reflex-dev/reflex/pull/3888

### Var Shadowing Detection

When a dynamic route arg name conflicts with the name of an existing state Var, the framework will raise `DynamicRouteArgShadowsStateVar`. This is intended to avoid unexpected behavior wherein the dynamic var takes precedence over the state var, even in a substate. A fix for the underlying cause is forthcoming.

In the meantime, define any pages that use dynamic route arguments _early_, so the dynamic var is available for use in the page functions.

* Dynamic route vars silently shadow all other vars by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3805
* cleanup dead test code by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3909

-----------------

When a computed var name conflicts with the name of an existing state Var, the framework will raise `ComputedVarShadowsStateVar`. ⚠️ [ENG-3758]

* Forbid Computed var shadowing by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3843
* remove reference to computed var by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3906

### `rx.progress` is now Radix Themes Progress

The previous Radis Primitives progress is available as `rx.radix.primitives.progress` ⚠️ [ENG-3756]

* [REF-3592]Promote `rx.progress` from radix themes by @ElijahAhianyo in https://github.com/reflex-dev/reflex/pull/3878

### Watchfiles dependency removed

If you are upgrading reflex in an existing environment run `pip uninstall watchfiles` first to ensure the hot reload mechanism is properly ignoring the `.web` directory.

### `DiskStateManager` - preserve state across hot reloads in dev mode
* implement disk state manager by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3826
* Disk state manager don't use byref by @picklelo in https://github.com/reflex-dev/reflex/pull/3874
* delete states if it exists on run by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3901

### `/_health` endpoint

```
http://localhost:8000/_health
```

```json
{"status":true,"db":true,"redis":false}
```

* /health endpoint for K8 Liveness and Readiness probes by @samarth9201 in https://github.com/reflex-dev/reflex/pull/3855

## Improvements

* improve state hierarchy validation, drop old testing special case by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3894
* fix var dependency dicts by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3842
* Update docker-example by @masenf in https://github.com/reflex-dev/reflex/pull/3324

## Var System

* fully migrate vars into new system by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3743
* guess_type: if the type is optional, treat it like it's "not None" by @masenf in https://github.com/reflex-dev/reflex/pull/3839
* Fix double-quoting of defaultColorMode by @masenf in https://github.com/reflex-dev/reflex/pull/3840
* ImmutableVar perf optimizations by @masenf in https://github.com/reflex-dev/reflex/pull/3814
* Get attribute access type fix by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3803
* fix var in bare by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3873
* add var_operation and move some operations to the new style by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3841
* Adding array to array pluck operation. by @abulvenz in https://github.com/reflex-dev/reflex/pull/3868
* simplify ImmutableComputedVar.__get__ by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3902
* Use old serializer system in LiteralVar by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3875
* Fix type propagation in ToStringOperation by @abulvenz in https://github.com/reflex-dev/reflex/pull/3895

## Documentation / Error Messages
* Add comments to DataList components by @elviskahoro in https://github.com/reflex-dev/reflex/pull/3827
* [REF-3589] raise EventHandlerArgMismatch when event handler args don't match spec by @masenf in https://github.com/reflex-dev/reflex/pull/3853

## Other Changes
* Fix benchmark tests by @ElijahAhianyo in https://github.com/reflex-dev/reflex/pull/3822
* [REF-3416] [REF-3632] Update fastapi, gunicorn, watchdog deps by @masenf in https://github.com/reflex-dev/reflex/pull/3859
* Remove watchfiles and watchdog [REF-2133] [REF-3694] by @picklelo in https://github.com/reflex-dev/reflex/pull/3862

## Bug Fixes

* [REF-3597]Type check Radio items by @ElijahAhianyo in https://github.com/reflex-dev/reflex/pull/3856
* Bug fix rx.toast. Allow passing `title` kwarg  by @TimChild in https://github.com/reflex-dev/reflex/pull/3857
* Update find_replace by @samarth9201 in https://github.com/reflex-dev/reflex/pull/3886
* Use correct flexgen backend URL by @masenf in https://github.com/reflex-dev/reflex/pull/3891
* gitignore .web by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3885
* [bugfix] Allow scroll in text_area when limited by max_height by @VishDev12 in https://github.com/reflex-dev/reflex/pull/3882
* Retain mutability inside `async with self` block by @masenf in https://github.com/reflex-dev/reflex/pull/3884
* Include child imports in markdown component_map by @masenf in https://github.com/reflex-dev/reflex/pull/3883
* mixin computed vars should only be applied to highest level state in … by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3833
* fix initial state without cv fallback by @benedikt-bartscher in https://github.com/reflex-dev/reflex/pull/3670
* add fragment to foreach by @adhami3310 in https://github.com/reflex-dev/reflex/pull/3877

## New Contributors
* @elviskahoro made their first contribution in https://github.com/reflex-dev/reflex/pull/3827
* @TimChild made their first contribution in https://github.com/reflex-dev/reflex/pull/3857
* @samarth9201 made their first contribution in https://github.com/reflex-dev/reflex/pull/3855
* @VishDev12 made their first contribution in https://github.com/reflex-dev/reflex/pull/3882

**Full Changelog**: https://github.com/reflex-dev/reflex/compare/v0.5.10...reflex-0.6.0

**Comment by elviskahoro** [comment]
> @elviskahoro made their first contribution in https://github.com/reflex-dev/reflex/pull/3827

First of many!

**Comment by dentroai** [comment]
Tested it out on our chatapp that supports image upload and stores messages in the inbuilt reflex database. 
Third party components are reflex-clerk and reflex-motion.

All seems to work fine!

**Comment by jq6l43d1** [comment]
I've tested 0.6.0a2 on my website and overall it's working great. The only wrapped component that doesn't work yet is [reflex-audio-capture](https://github.com/masenf/reflex-audio-capture).

**Reply by masenf** [reply]
i just published a new version of reflex-audio-capture v0.0.6 that should be good with reflex v0.6.0a3

**Comment by masenf** [comment]
It's out! https://github.com/reflex-dev/reflex/releases/tag/v0.6.0 Go use it!

**Reply by Ernelene** [reply]
Hi I have quite a big app with nested states to keep things a bit more organised. I am having a problem with the File name being too long. 
Any help would be appreciated.

`File "/Users/.../Documents/Code/.../reflex_frontend/.venv/lib/python3.12/site-packages/reflex/state.py", line 2779, in set_state_for_substate
    await self.set_state_for_substate(client_token, substate_substate)
  File "/Users/.../Documents/Code/.../reflex_frontend/.venv/lib/python3.12/site-packages/reflex/state.py", line 2776, in set_state_for_substate
    self.token_path(substate_token).write_bytes(state_dilled)
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pathlib.py", line 1037, in write_bytes
    with self.open(mode='wb') as f:
         ^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/pathlib.py", line 1013, in open
    return io.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 63] File name too long: '/Users/.../Documents/Code/.../reflex_frontend/.web/states/83783768-0869-4424-cd2c-4cbabd5a11a5_reflex___state____state.frontend___pages___global_state____global_state.frontend___pages___setup___workspaces___workspace_state____workspace_state.frontend___pages___setup___workspaces___workspace_components___drawer___drawer_state____drawer_state.pkl'`

**Reply by adhami3310** [reply]
a PR for this is up: https://github.com/reflex-dev/reflex/pull/4000

> I am having a problem with the File name being too long.
